libsnark/zk_proof_systems/plonk/common_input.tcc

```diff
@@ -0,0 +1,138 @@
+/** @file
+*****************************************************************************
+Implementation of Common Preprocessed Input interfaces for a ppzkSNARK
+for Plonk.
+
+See common_input.hpp .
+
+*****************************************************************************
+* @author     This file is part of libsnark, developed by SCIPR Lab
+*             and contributors (see AUTHORS).
+* @copyright  MIT license (see LICENSE file)
+*****************************************************************************/
+
+#ifndef PLONK_PPZKSNARK_COMMON_INPUT_TCC_
+#define PLONK_PPZKSNARK_COMMON_INPUT_TCC_
+
+namespace libsnark
+{
+template<typename ppT>
+void common_preprocessed_input<ppT>::setup_from_example(
+    plonk_example<ppT> example)
+{
+    using Field = libff::Fr<ppT>;
+
+    // public input (PI)
+    Field PI_value = example.public_input;
+    // index of the row of the PI in the non-transposed gates_matrix
+    int PI_index = example.public_input_index;
+    // Transposed gates matrix: each row is a q-vector
+    std::vector<std::vector<Field>> gates_matrix_transpose =
+        example.gates_matrix_transpose;
+    // wire permutation
+    std::vector<size_t> wire_permutation = example.wire_permutation;
+    // Generate domains on which to evaluate the witness
+    // polynomials. k1,k2 can be random, but we fix them for debug to
+    // match against the test vector values
+    Field k1 = example.k1;
+    Field k2 = example.k2;
+    this->k1 = example.k1;
+    this->k2 = example.k2;
+#ifdef DEBUG
+    printf("[%s:%d] k1 ", __FILE__, __LINE__);
+    k1.print();
+    printf("[%s:%d] k2 ", __FILE__, __LINE__);
+    k2.print();
+#endif // #ifdef DEBUG
+
+    this->num_gates = example.num_gates;
+#ifdef DEBUG
+    // ensure num_gates is power of 2
+    bool b_is_power2 = ((this->num_gates & (this->num_gates - 1)) == 0);
+    assert(b_is_power2);
+    // ensure that num_gates is not 0
+    assert(this->num_gates);
```

Comment:
asserts only check the passed expression in debug mode, so this one may be moved directly under `this->num_gates = example.num_gates;`

Reply:
Done.